cmd/fs-v1-multipart.go

```diff
@@ -983,80 +646,67 @@ func (fs fsObjects) CompleteMultipartUpload(bucket string, object string, upload
-		return err
-	}
-
-	defer func() {
-		if removeObjectDir {
-			basePath := pathJoin(fs.fsPath, minioMetaMultipartBucket, bucket)
-			derr := fsDeleteFile(basePath, pathJoin(basePath, object))
-			if derr = errors.Cause(derr); derr != nil {
-				// In parallel execution, AbortMultipartUpload could have deleted temporary
-				// state files/directory, it is safe to ignore errFileNotFound
-				if derr != errFileNotFound {
-					errorIf(derr, "unable to remove %s in %s", pathJoin(basePath, object), basePath)
-				}
-			}
-		}
-		objectMPartPathLock.Unlock()
-	}()
-
-	fsMetaPath := pathJoin(fs.fsPath, minioMetaMultipartBucket, uploadIDPath, fsMetaJSONFile)
-	if _, err := fs.rwPool.Open(fsMetaPath); err != nil {
-		if err == errFileNotFound || err == errFileAccessDenied {
+	uploadIDDir := fs.getUploadIDDir(bucket, object, uploadID)
+	// Just check if the uploadID exists to avoid copy if it doesn't.
+	_, err := fsStatFile(pathJoin(uploadIDDir, fsMetaJSONFile))
+	if err != nil {
+		if errors.Cause(err) == errFileNotFound || errors.Cause(err) == errFileAccessDenied {
 			return errors.Trace(InvalidUploadID{UploadID: uploadID})
 		}
 		return toObjectErr(errors.Trace(err), bucket, object)
 	}
 
-	uploadsPath := pathJoin(bucket, object, uploadsJSONFile)
-	rwlk, err := fs.rwPool.Write(pathJoin(fs.fsPath, minioMetaMultipartBucket, uploadsPath))
-	if err != nil {
-		fs.rwPool.Close(fsMetaPath)
-		if err == errFileNotFound || err == errFileAccessDenied {
-			return errors.Trace(InvalidUploadID{UploadID: uploadID})
-		}
+	if err := fsRemoveAll(uploadIDDir); err != nil {
 		return toObjectErr(errors.Trace(err), bucket, object)
 	}
-	defer rwlk.Close()
 
-	// Signal appendParts routine to stop waiting for new parts to arrive.
-	fs.abort(uploadID)
-
-	// Close lock held on bucket/object/uploadid/fs.json,
-	// this needs to be done for windows so that we can happily
-	// delete the bucket/object/uploadid
-	fs.rwPool.Close(fsMetaPath)
-
-	// Cleanup all uploaded parts and abort the upload.
-	multipartObjectDir := pathJoin(fs.fsPath, minioMetaMultipartBucket, bucket, object)
-	multipartUploadIDDir := pathJoin(multipartObjectDir, uploadID)
-	if err = fsRemoveUploadIDPath(multipartObjectDir, multipartUploadIDDir); err != nil {
-		return toObjectErr(err, bucket, object)
-	}
+	return nil
+}
 
-	// Remove entry from `uploads.json`.
-	removeObjectDir, err = fs.removeUploadID(bucket, object, uploadID, rwlk)
-	if err != nil {
-		return toObjectErr(err, bucket, object)
+// Removes multipart uploads if any older than `expiry` duration
+// on all buckets for every `cleanupInterval`, this function is
+// blocking and should be run in a go-routine.
+func (fs *fsObjects) cleanupStaleMultipartUploads(cleanupInterval, expiry time.Duration, doneCh chan struct{}) {
+	ticker := time.NewTicker(cleanupInterval)
+	for {
+		select {
+		case <-doneCh:
+			// Stop the timer.
+			ticker.Stop()
+			return
+		case <-ticker.C:
+			now := time.Now()
```

Comment:
do you not want to use `UTCNow()` ?

Reply:
no because fi.Modtime() (which we use to compare with `now`) does not return in UTC

Reply:
Are you sure ?

Reply:
check this:
```
krishna@escape:~/dev/goprogs$ cat mtime.go
package main

import (
	"fmt"
	"log"
	"os"
	"time"
)

func main() {
	fmt.Println(time.Now(), time.Now().UTC())
	fi, err := os.Stat("mtime.go")
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println(fi.ModTime())
}
krishna@escape:~/dev/goprogs$ touch mtime.go
krishna@escape:~/dev/goprogs$ go run mtime.go
2018-01-17 12:44:04.361967289 -0800 PST m=+0.000160832 2018-01-17 20:44:04.361967415 +0000 UTC
2018-01-17 12:44:00.791171574 -0800 PST
krishna@escape:~/dev/goprogs$ 
```

Reply:
Actually, looks like it does not matter as time package abstracts it:
```
krishna@escape:~/dev/goprogs$ cat mtime.go
package main

import (
	"fmt"
	"log"
	"os"
	"time"
)

func main() {
	now := time.Now()
	nowUTC := time.Now().UTC()
	fi, err := os.Stat("mtime.go")
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println(now.Sub(fi.ModTime()), nowUTC.Sub(fi.ModTime()))
}
krishna@escape:~/dev/goprogs$ go run mtime.go
44.118449376s 44.118449493s
krishna@escape:~/dev/goprogs$ 
```

Reply:
So they manage it for us..